Docs/dlt plus project docs rest api restructuring (#2911)
dlt-hub/dlt: docs/website/docs/plus/features/project/source-configuration.md

---

> title: Configuration of Source

Reviewer: ```suggestion
title: Source configuration
```

> description: Initialize ore sources in yaml file

Reviewer: ```suggestion
description: How to setup sources in YAML file
```

---

> # Configuration of Sources

Reviewer: ```suggestion
# Source configuration
```

> <img src="https://storage.googleapis.com/dlt-blog-images/plus/dlt_plus_projects.png" width="500"/>

Reviewer: I think image in the overview is enoug

> The `dlt.yml` file is used to declaratively configure data sources for use with the dlt CLI. This setup provides a flexible way to manage data extraction for the core sources REST APIs, SQL databases and file systems using YAML syntax.

Reviewer: I'm not sure I've got the first sentence, maybe something along the lines:
```suggestion
The dlt.yml file enables a fully declarative setup of your data source and its parameters. It supports built-in sources such as REST APIs, SQL databases, and cloud storage, as well as any custom source you define.
```

> Credential placeholders for the defined sources are automatically generated in `.dlt/secrets.toml`. Alternatively, credentials may also be provided directly within `dlt.yml`.

Reviewer: We do not recommend it :c 
Let's say the configuration may be provided in `dlt.yml`

Author: Not sure if I understand what you mean.. Like this: "Alternatively, configuration may also be provided directly within `dlt.yml`." ?

## REST API

> The built-in `rest_api` type enables configuration of REST-based integrations. Multiple endpoints can be defined under a single source.

Reviewer: Let's add the link to res api source documentation, where people can find details on all available configuration https://deploy-preview-2925--dlt-hub-docs.netlify.app/docs/devel/dlt-ecosystem/verified-sources/rest_api/basic

```yaml
sources:
  pokemon_api:
    type: rest_api
    client:
      base_url: https://pokeapi.co/api/v2/
      paginator: auto
    resource_defaults:
      primary_key: name
    resources:
      - pokemon
      - berry
      - 

>         name: encounter_conditions

Reviewer: ```suggestion
        name: encounter_condition
```

        endpoint:
          path: encounter-condition
          params:
            offset:
              type: incremental
              cursor_path: name
        write_disposition: append
```

* `type: rest_api`: Specifies the use of the built-in REST API source.
* `client.base_url`: Sets the root URL for all API requests.
* `paginator: auto`: Enables automatic detection and handling of pagination.
* `resource_defaults`: Contains the default values to configure the dlt resources. This configuration is applied to all resources unless overridden by the resource-specific configuration.

> * Each item in `resources`defines an endpoint to extract. Simple entries like `pokemon`and `berry` will fetch from `/pokemon` and `/berry`, respectively.

Reviewer: ```suggestion
* Each item in `resources` defines an endpoint to extract. Simple entries like `pokemon` and `berry` will fetch from `/pokemon` and `/berry`, respectively.
```

* The `encounter-condition` resource uses an advanced configuration:

>   * `path`: Point to the `/encounter-condition`endpoint.

Reviewer: ```suggestion
  * `path`: Point to the `/encounter-condition` endpoint.
```

  * `params.offset`: Enables incremental loading using the `name` field as the cursor.

>   * `write_disposition: append`: Ensures new data is appended rather than overwriting previous loads.

Reviewer: Maybe let's use any other write disposition, because `append` is the default one, so it might be confusing why we're specifying it

## SQL database

> For SQL-base extractions that require no table-specific parameter configuration, it's possible to initialize `type = sql_database` and declare multiple tables at once.

Reviewer: ```suggestion
For SQL-base extractions that require no table-specific parameter configuration, it's possible to initialize `type: sql_database` and declare multiple tables at once.
```

 
### General SQL Database Source

```yaml 
sources:
   sql_source:
    type: sql_database
    table_names: 
      - family
      - clan

>     incremental:

Reviewer: Is it working with `sql_database` ?

Author: Yes.

        cursor_path: updated
        initial_value: 2023-01-12T11:21:28Z 
```

This defines a connection to a SQL database with incremental loading applied across multiple tables.

* `type: sql_database`: Specifies the SQL database connector.

* `table_names`: List of tables to extract.

* `incremental`: Global configuration for incremental extraction.

### Table-Specific Configuration

For table specific configurationssettings such as different `primary_key`s, individual tables can be defined as standalone sources using the `sql_table` type.

```yaml
sources: 
  sql_family:

>       type: sql_database.sql_table

Reviewer: Will `type: sql_table` work?

      table: family
      incremental:
        cursor_path: updated
        initial_value: 2023-01-12T11:21:28Z 
      primary_key: rfam_id
```

* `type: sql_table`: Indicates a single-table extraction.

* `table`: Name of the table to extract.

* `incremental`: Enables incremental loading for the table.

* `primary_key`: Specifies the table's unique identifier for deduplication and merges.

## Filesystem

Filesystem sources can be set via the `readers` type and the filesystem specific resources can be called via the CLI `run pipline` command.

```yaml
sources: 
 file_source:

>     type: filesystem.readers

Reviewer: Shouldn't it be 
```suggestion
    type: filesystem.readers.read_csv
```
?
Makes sense to define it here once instead of calling the --resources on each run
